corehq/apps/users/models.py

```diff
@@ -2334,6 +2334,11 @@ def get_location_map_case(self):
     def memoized_usercase(self):
         return self.get_usercase()
 
+    def get_usercase_by_domain(self, domain):
```

Comment:
nit: I prefer how it was before, because there isn't a concept of "by_domain" for a CommCare User doesn't make sense since they only belong to one domain. Also, returning None if the passed in domain doesn't match just seems unexpected.

Reply:
I think this is to make it so usercases can be accessed the same way whether the user is a web user or mobile worker.  Looking at the code, I see the following accessors:

* CommCareUser
  * memoized_usercase
  * get_usercase
  * get_usercase_id
  * get_usercase_by_domain [new]
* WebUser
  * get_usercase_id
  * get_usercase_by_domain

I see `get_usercase` and `get_usercase_by_domain` have the same implementation, and the other methods all reference one of those two, so it seems to me we could move everything to `CouchUser`.  Looking briefly at usages, I see `get_usercase_id` is only used on restore users, and could then be moved to the parent there too (and probably moved in-line).  `get_usercase` isn't used anywhere other than these other methods, nor is `get_usercase_by_domain`.  In fact, it looks like only `memoized_usercase` is used anywhere else.  I think this could all be reduced to a single method - I'd be happy to do a cleanup pass myself after this is merged.

(Aside: I prefer `get_usercase` as a name, if we're picking one - "by domain" is implied if that's a required arg)

Reply:
If you move it into CouchUser, won't you have to make the check what kind of user it is explicitly? I can see adding an unimplemented method there but I don't see why the type system should not choose implementation.